docs/reference/speed_python.rst

```diff
@@ -0,0 +1,311 @@
+can take several milliseconds.
+
+Consequently the performance of a function or method can be improved if an object is created
+once only and not permitted to grow in size. This implies that the object persists
+for the duration of its use: typically it will be instantiated in a class constructor
+and used in various methods.
+
+This is covered in further detail :ref:`Controlling garbage collection <gc>` below.
+
+Buffers
+~~~~~~~
+
+An example of the above is the common case where a buffer is required, such as one
+used for communication with a device. A typical driver will create the buffer in the
+constructor and use it in its I/O methods which will be called repeatedly.
+
+The MicroPython libraries typically provide optional support pre-allocated buffers.
+For example the ``uart.readinto()`` method allows two options for its argument, an integer
+or a buffer. If an integer is supplied it will read up to that number of bytes and
+return the outcome: this implies that a buffer is created with a corresponding
+memory allocation. Providing a pre-allocated buffer as the argument avoids this. See
+the code fragment in :ref:`Cacheing object references <Cacheing>` below.
+
+Floating Point
+~~~~~~~~~~~~~~
+
+For the most speed critical sections of code it is worth noting that performing
+any kind of floating point operation involves heap allocation. Where possible use
+integer operations and restrict the use of floating point to sections of the code
+where performance is not paramount.
+
+Arrays
+~~~~~~
+
+Python lists are convenient but randomly accessing list elements is relatively slow
+involving navigation through a linked list. The ``array`` module provides an alternative 
+along with Python's built in ``bytes`` and ``bytearray`` classes. These data
+structures store elements in contiguous memory locations enabling fast random
+access. Once again to avoid memory allocation in critical code these
+should be pre-allocated and passed as arguments or as bound objects.
+
+When passing slices of objects such as ``bytearray`` instances, Python creates
+a copy which involves allocation. This can be avoided using a ``memoryview``
+object:
+
+.. code:: python
+
+    ba = bytearray(100)
+    func(ba[3:10]) # a copy is passed
+    mv = memoryview(ba)
+    func(mv[3:10]) # a pointer to memory is passed
+
+The ``memoryview`` can only be applied to objects whose data is contained in contiguous
+memory locations (objects supporting the buffer protocol) - another reason for using
+arrays rather than lists.
+
+Identifying the slowest section of code
+---------------------------------------
+
+This is a process known as profiling and is covered in textbooks and
+(for standard Python) supported by various software tools. For the type of
+smaller embedded application likely to be running on MicroPython platforms
+the slowest function or method can usually be established by judicious use
+of ``print(pyb.elapsed_micros(start))`` statements.
```

Comment:
Note that pyboard now has time.ticks_XXX functions which are recommended over the (old) pyb.millis and pyb.elapsed_millis (and micros versions) functions.  There is also time.ticks_cpu() which can count actualy CPU cycles.